File: app/epub.py
from __future__ import annotations

import posixpath
import re
from dataclasses import dataclass
from html import unescape
from html.parser import HTMLParser
from pathlib import Path
from xml.etree import ElementTree as ET
from zipfile import ZipFile


@dataclass(slots=True)
class ExtractedBook:
    title: str
    author: str
    paragraphs: list[str]
    chapters: list["ExtractedChapter"]


@dataclass(slots=True)
class ExtractedChapter:
    title: str
    source_href: str
    start_paragraph_index: int
    end_paragraph_index: int


@dataclass(slots=True)
class ExtractedChapterBlock:
    kind: str
    text: str = ""
    image_src: str = ""
    alt: str = ""


class EpubTextParser(HTMLParser):
    block_tags = {"p", "div", "br", "h1", "h2", "h3", "h4", "li"}

    def __init__(self) -> None:
        super().__init__()
        self.out: list[str] = []

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        attr = dict(attrs)
        if tag == "img" and "Art_orn" in attr.get("src", ""):
            self.out.append("\n\nX X X\n\n")
        elif tag in self.block_tags:
            self.out.append("\n")

    def handle_endtag(self, tag: str) -> None:
        if tag in self.block_tags:
            self.out.append("\n")

    def handle_data(self, data: str) -> None:
        self.out.append(data)

    def text(self) -> str:
        text = unescape("".join(self.out))
        text = re.sub(r"[ \t]+", " ", text)
        return re.sub(r"\n{3,}", "\n\n", text).strip()


def split_paragraphs(text: str) -> list[str]:
    return [paragraph.strip() for paragraph in text.split("\n\n") if paragraph.strip()]


def parse_chapter_blocks(content: str, source_href: str) -> list[ExtractedChapterBlock]:
    parser = EpubContentParser(source_href)
    parser.feed(content)
    return parser.blocks()


def read_epub(path: Path) -> ExtractedBook:
    with ZipFile(path) as book:
        container = ET.fromstring(book.read("META-INF/container.xml"))
        opf_path = container.find(".//{*}rootfile").attrib["full-path"]
        opf = ET.fromstring(book.read(opf_path))
        manifest = {
            item.attrib["id"]: item.attrib["href"]
            for item in opf.findall(".//{*}manifest/{*}item")
        }
        metadata = opf.find(".//{*}metadata")
        title = _first_metadata_text(metadata, "title") or path.stem
        author = _first_metadata_text(metadata, "creator") or ""
        base = posixpath.dirname(opf_path)
        chapters: list[ExtractedChapter] = []
        paragraphs: list[str] = []
        chapter_titles = _read_toc_labels(book, opf, base)
        for itemref in opf.findall(".//{*}spine/{*}itemref"):
            href = manifest.get(itemref.attrib["idref"], "")
            if not href.endswith((".xhtml", ".html")):
                continue
            source_href = posixpath.normpath(href)
            content = book.read(posixpath.normpath(posixpath.join(base, href))).decode("utf-8", "ignore")
            chapter_blocks = parse_chapter_blocks(content, source_href)
            if not chapter_blocks:
                continue
            chapter_paragraphs = [block.text for block in chapter_blocks if block.kind == "paragraph"]
            start = len(paragraphs)
            paragraphs.extend(chapter_paragraphs)
            chapters.append(
                ExtractedChapter(
                    title=chapter_titles.get(source_href) or _fallback_chapter_title(source_href, len(chapters)),
                    source_href=source_href,
                    start_paragraph_index=start,
                    end_paragraph_index=len(paragraphs) - 1,
                )
            )

    return ExtractedBook(title=title.strip() or path.stem, author=author.strip(), paragraphs=paragraphs, chapters=chapters)


def read_epub_chapter_blocks(path: Path, source_href: str) -> list[ExtractedChapterBlock]:
    with ZipFile(path) as book:
        container = ET.fromstring(book.read("META-INF/container.xml"))
        opf_path = container.find(".//{*}rootfile").attrib["full-path"]
        base = posixpath.dirname(opf_path)
        full_path = posixpath.normpath(posixpath.join(base, source_href))
        content = book.read(full_path).decode("utf-8", "ignore")
        return parse_chapter_blocks(content, source_href)


def read_epub_asset(path: Path, source_href: str, asset_href: str) -> tuple[bytes, str]:
    with ZipFile(path) as book:
        container = ET.fromstring(book.read("META-INF/container.xml"))
        opf_path = container.find(".//{*}rootfile").attrib["full-path"]
        base = posixpath.dirname(opf_path)
        asset_path = posixpath.normpath(posixpath.join(base, posixpath.dirname(source_href), asset_href))
        return book.read(asset_path), asset_path


def slugify(value: str) -> str:
    slug = value.lower().replace("volume", "vol")
    slug = re.sub(r"[^a-z0-9]+", "-", slug)
    return re.sub(r"-+", "-", slug).strip("-") or "book"


def _first_metadata_text(metadata: ET.Element | None, name: str) -> str:
    if metadata is None:
        return ""
    for child in metadata:
        if child.tag.endswith(name) and child.text:
            return child.text
    return ""


def _read_toc_labels(book: ZipFile, opf: ET.Element, base: str) -> dict[str, str]:
    labels: dict[str, str] = {}
    manifest_nodes = opf.findall(".//{*}manifest/{*}item")
    ncx_href = next((item.attrib.get("href", "") for item in manifest_nodes if item.attrib.get("media-type") == "application/x-dtbncx+xml"), "")
    if not ncx_href:
        return labels
    try:
        toc = ET.fromstring(book.read(posixpath.normpath(posixpath.join(base, ncx_href))))
    except KeyError:
        return labels
    for nav in toc.findall(".//{*}navPoint"):
        content = nav.find(".//{*}content")
        label = nav.find(".//{*}text")
        if content is None or label is None or not label.text:
            continue
        source_href = content.attrib.get("src", "").split("#", 1)[0]
        if not source_href:
            continue
        labels[posixpath.normpath(source_href)] = " ".join(label.text.split())
    return labels


def _fallback_chapter_title(source_href: str, chapter_index: int) -> str:
    stem = Path(source_href).stem.replace("_", " ").replace("-", " ").strip()
    return stem.title() if stem else f"Chapter {chapter_index + 1}"


class EpubContentParser(HTMLParser):
    paragraph_tags = {"p", "div", "li", "h1", "h2", "h3", "h4"}

    def __init__(self, source_href: str) -> None:
        super().__init__()
        self.source_href = source_href
        self.current_tag: str | None = None
        self.current_text: list[str] = []
        self.current_href: str | None = None
        self.current_img_alt: str = ""
        self._blocks: list[ExtractedChapterBlock] = []

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        attr = dict(attrs)
        if tag == "img":
            src = attr.get("src", "")
            if "Art_orn" in src:
                self._blocks.append(
                    ExtractedChapterBlock(
                        kind="image",
                        image_src=posixpath.normpath(src),
                        alt=(attr.get("alt") or "").strip(),
                    )
                )
                return
            inline_symbol = inline_image_symbol(src, attr.get("class", ""), attr.get("style", ""))
            if inline_symbol is not None:
                self.current_text.append(inline_symbol)
                return
            if src:
                self._blocks.append(
                    ExtractedChapterBlock(
                        kind="image",
                        image_src=posixpath.normpath(src),
                        alt=(attr.get("alt") or "").strip(),
                    )
                )
            return
        if tag == "a":
            self.current_href = attr.get("href")
        if tag in self.paragraph_tags:
            self._flush_text()
            self.current_tag = tag
            self.current_text = []
            return
        if tag == "br":
            self.current_text.append("\n")

    def handle_endtag(self, tag: str) -> None:
        if tag == "a":
            self.current_href = None
        if tag == self.current_tag:
            self._flush_text()
            self.current_tag = None

    def handle_data(self, data: str) -> None:
        if self.current_tag:
            self.current_text.append(data)

    def blocks(self) -> list[ExtractedChapterBlock]:
        self._flush_text()
        return self._blocks

    def _flush_text(self) -> None:
        if not self.current_text:
            return
        text = unescape("".join(self.current_text))
        text = re.sub(r"[ \t]+", " ", text)
        text = re.sub(r"\n{2,}", "\n", text).strip()
        self.current_text = []
        if not text:
            return
        if self.current_tag and self.current_tag.startswith("h") and self.current_href and "toc.xhtml" in self.current_href:
            return
        self._append_paragraph(text)

    def _append_paragraph(self, text: str) -> None:
        self._blocks.append(ExtractedChapterBlock(kind="paragraph", text=text))


def inline_image_symbol(src: str, class_name: str, style: str) -> str | None:
    normalized_src = posixpath.basename(src)
    normalized_class = class_name or ""
    normalized_style = style or ""
    if normalized_src == "Art_1.jpg" or "orn2" in normalized_class:
        return "⑧"
    if normalized_src in {"Art_star1.jpg", "Art_star2.jpg"} or "orn1" in normalized_class:
        return "★"
    if normalized_src == "Art_music.jpg":
        return "♪"
    if "vertical-align" in normalized_style:
        return ""
    return None

File: src-tauri/src/epub.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::Read;
use std::path::Path;

use anyhow::{anyhow, Context, Result};
use roxmltree::Document;
use scraper::{ElementRef, Html, Selector};
use zip::ZipArchive;

pub struct ExtractedBook {
    pub title: String,
    pub author: String,
    pub cover: Option<ExtractedCover>,
    pub chapters: Vec<ExtractedChapter>,
}

pub struct ExtractedCover {
    pub path: String,
    pub bytes: Vec<u8>,
}

pub struct ExtractedChapter {
    pub title: String,
    pub source_href: String,
    pub blocks: Vec<ExtractedBlock>,
}

pub struct ExtractedBlock {
    pub kind: ExtractedBlockKind,
    pub text: String,
    pub asset_path: Option<String>,
    pub alt: String,
}

#[derive(Clone, Copy)]
pub enum ExtractedBlockKind {
    Paragraph,
    Image,
}

#[derive(Clone)]
struct ManifestItem {
    id: String,
    href: String,
    media_type: String,
    properties: String,
}

pub fn read_epub(path: &Path) -> Result<ExtractedBook> {
    let file = File::open(path)?;
    let mut zip = ZipArchive::new(file)?;
    let container = read_zip_text(&mut zip, "META-INF/container.xml")?;
    let container_doc = Document::parse(&container)?;
    let opf_path = container_doc
        .descendants()
        .find(|node| node.tag_name().name() == "rootfile")
        .and_then(|node| node.attribute("full-path"))
        .ok_or_else(|| anyhow!("EPUB container is missing the OPF rootfile."))?
        .to_string();
    let opf = read_zip_text(&mut zip, &opf_path)?;
    let opf_doc = Document::parse(&opf)?;
    let opf_base = zip_dirname(&opf_path);

    let title = metadata_text(&opf_doc, "title")
        .filter(|value| !value.trim().is_empty())
        .unwrap_or_else(|| {
            path.file_stem()
                .and_then(|value| value.to_str())
                .unwrap_or("Book")
                .to_string()
        });
    let author = metadata_text(&opf_doc, "creator").unwrap_or_default();
    let manifest = read_manifest(&opf_doc);
    let manifest_by_id: HashMap<String, ManifestItem> = manifest
        .iter()
        .cloned()
        .map(|item| (item.id.clone(), item))
        .collect();
    let toc_labels = read_toc_labels(&mut zip, &opf_base, &manifest).unwrap_or_default();
    let cover = find_cover(&mut zip, &opf_base, &opf_doc, &manifest, &manifest_by_id);

    let mut chapters = Vec::new();
    for itemref in opf_doc
        .descendants()
        .filter(|node| node.tag_name().name() == "itemref")
    {
        let Some(idref) = itemref.attribute("idref") else {
            continue;
        };
        let Some(item) = manifest_by_id.get(idref) else {
            continue;
        };
        if !looks_like_html(&item.href, &item.media_type) {
            continue;
        }
        let source_href = normalize_zip_path(&item.href);
        let zip_path = join_zip_path(&opf_base, &item.href);
        let content = read_zip_text(&mut zip, &zip_path)
            .with_context(|| format!("Unable to read chapter {}", item.href))?;
        let blocks = parse_chapter_blocks(&content, &zip_dirname(&zip_path));
        if blocks.is_empty() {
            continue;
        }
        let title = toc_labels
            .get(&source_href)
            .cloned()
            .unwrap_or_else(|| fallback_chapter_title(&source_href, chapters.len()));
        chapters.push(ExtractedChapter {
            title,
            source_href,
            blocks,
        });
    }

    Ok(ExtractedBook {
        title: normalize_text(&title).unwrap_or_else(|| "Book".to_string()),
        author: normalize_text(&author).unwrap_or_default(),
        cover,
        chapters,
    })
}

pub fn read_chapter_blocks(path: &Path, source_href: &str) -> Result<Vec<ExtractedBlock>> {
    let file = File::open(path)?;
    let mut zip = ZipArchive::new(file)?;
    let zip_path = find_zip_entry(&mut zip, source_href)
        .ok_or_else(|| anyhow!("Unable to find chapter {source_href} in EPUB."))?;
    let content = read_zip_text(&mut zip, &zip_path)?;
    Ok(parse_chapter_blocks(&content, &zip_dirname(&zip_path)))
}

pub fn read_asset_bytes(path: &Path, asset_path: &str) -> Result<Vec<u8>> {
    let file = File::open(path)?;
    let mut zip = ZipArchive::new(file)?;
    let zip_path = find_zip_entry(&mut zip, asset_path)
        .ok_or_else(|| anyhow!("Unable to find asset {asset_path} in EPUB."))?;
    read_zip_bytes(&mut zip, &zip_path)
}

fn read_zip_text(zip: &mut ZipArchive<File>, path: &str) -> Result<String> {
    let mut file = zip.by_name(path)?;
    let mut content = String::new();
    file.read_to_string(&mut content)?;
    Ok(content)
}

fn read_zip_bytes(zip: &mut ZipArchive<File>, path: &str) -> Result<Vec<u8>> {
    let mut file = zip.by_name(path)?;
    let mut content = Vec::new();
    file.read_to_end(&mut content)?;
    Ok(content)
}

fn find_zip_entry(zip: &mut ZipArchive<File>, source_href: &str) -> Option<String> {
    let normalized = normalize_zip_path(source_href);
    zip.file_names()
        .find(|path| {
            let candidate = normalize_zip_path(path);
            candidate == normalized || candidate.ends_with(&format!("/{normalized}"))
        })
        .map(ToString::to_string)
}

fn metadata_text(doc: &Document<'_>, name: &str) -> Option<String> {
    doc.descendants()
        .find(|node| node.tag_name().name() == name)
        .and_then(|node| node.text())
        .and_then(normalize_text)
}

fn read_manifest(doc: &Document<'_>) -> Vec<ManifestItem> {
    doc.descendants()
        .filter(|node| node.tag_name().name() == "item")
        .map(|node| ManifestItem {
            id: node.attribute("id").unwrap_or_default().to_string(),
            href: node.attribute("href").unwrap_or_default().to_string(),
            media_type: node.attribute("media-type").unwrap_or_default().to_string(),
            properties: node.attribute("properties").unwrap_or_default().to_string(),
        })
        .filter(|item| !item.id.is_empty() && !item.href.is_empty())
        .collect()
}

fn read_toc_labels(
    zip: &mut ZipArchive<File>,
    opf_base: &str,
    manifest: &[ManifestItem],
) -> Result<HashMap<String, String>> {
    let Some(ncx) = manifest
        .iter()
        .find(|item| item.media_type == "application/x-dtbncx+xml")
    else {
        return Ok(HashMap::new());
    };
    let toc_path = join_zip_path(opf_base, &ncx.href);
    let toc_base = zip_dirname(&toc_path);
    let content = read_zip_text(zip, &toc_path)?;
    let doc = Document::parse(&content)?;
    let mut labels = HashMap::new();
    for nav_point in doc
        .descendants()
        .filter(|node| node.tag_name().name() == "navPoint")
    {
        let label = nav_point
            .descendants()
            .find(|node| node.tag_name().name() == "text")
            .and_then(|node| node.text())
            .and_then(normalize_text);
        let src = nav_point
            .descendants()
            .find(|node| node.tag_name().name() == "content")
            .and_then(|node| node.attribute("src"));
        if let (Some(label), Some(src)) = (label, src) {
            let no_fragment = src.split('#').next().unwrap_or(src);
            let absolute = join_zip_path(&toc_base, no_fragment);
            let source_href = strip_base(&absolute, opf_base);
            labels.insert(source_href, label);
        }
    }
    Ok(labels)
}

fn find_cover(
    zip: &mut ZipArchive<File>,
    opf_base: &str,
    doc: &Document<'_>,
    manifest: &[ManifestItem],
    manifest_by_id: &HashMap<String, ManifestItem>,
) -> Option<ExtractedCover> {
    let metadata_cover_id = doc
        .descendants()
        .find(|node| {
            node.tag_name().name() == "meta"
                && node
                    .attribute("name")
                    .is_some_and(|value| value.eq_ignore_ascii_case("cover"))
        })
        .and_then(|node| node.attribute("content"));
    let candidates = [
        metadata_cover_id.and_then(|id| manifest_by_id.get(id)),
        manifest.iter().find(|item| {
            item.properties
                .split_whitespace()
                .any(|value| value == "cover-image")
        }),
        manifest.iter().find(|item| {
            item.href.to_lowercase().contains("cover") && item.media_type.starts_with("image/")
        }),
        manifest
            .iter()
            .find(|item| item.media_type.starts_with("image/")),
    ];

    for candidate in candidates.into_iter().flatten() {
        let path = join_zip_path(opf_base, &candidate.href);
        if let Ok(bytes) = read_zip_bytes(zip, &path) {
            return Some(ExtractedCover { path, bytes });
        }
    }
    None
}

fn parse_chapter_blocks(content: &str, chapter_base: &str) -> Vec<ExtractedBlock> {
    let document = Html::parse_document(content);
    let selector =
        Selector::parse("p, div, li, h1, h2, h3, h4, img").expect("valid chapter block selector");
    document
        .select(&selector)
        .filter(|element| !is_redundant_container(element))
        .filter_map(|element| parse_chapter_block(element, chapter_base))
        .collect()
}

fn parse_chapter_block(element: ElementRef<'_>, chapter_base: &str) -> Option<ExtractedBlock> {
    if element.value().name() == "img" {
        let source = element.value().attr("src")?;
        return Some(ExtractedBlock {
            kind: ExtractedBlockKind::Image,
            text: String::new(),
            asset_path: Some(join_zip_path(chapter_base, source)),
            alt: element
                .value()
                .attr("alt")
                .unwrap_or_default()
                .trim()
                .to_string(),
        });
    }

    normalize_text(&element.text().collect::<Vec<_>>().join(" ")).map(|text| ExtractedBlock {
        kind: ExtractedBlockKind::Paragraph,
        text,
        asset_path: None,
        alt: String::new(),
    })
}

fn is_redundant_container(element: &ElementRef<'_>) -> bool {
    if element.value().name() != "div" {
        return false;
    }
    has_readable_descendant(element)
}

fn has_readable_descendant(element: &ElementRef<'_>) -> bool {
    element.children().any(|child| {
        ElementRef::wrap(child).is_some_and(|child_element| {
            matches!(
                child_element.value().name(),
                "p" | "li" | "h1" | "h2" | "h3" | "h4" | "img"
            ) || has_readable_descendant(&child_element)
        })
    })
}

fn looks_like_html(href: &str, media_type: &str) -> bool {
    media_type.contains("html") || href.ends_with(".xhtml") || href.ends_with(".html")
}

fn normalize_text(value: &str) -> Option<String> {
    let normalized = value.split_whitespace().collect::<Vec<_>>().join(" ");
    if normalized.is_empty() {
        None
    } else {
        Some(normalized)
    }
}

fn fallback_chapter_title(source_href: &str, index: usize) -> String {
    let stem = source_href
        .rsplit('/')
        .next()
        .and_then(|value| value.split('.').next())
        .unwrap_or_default()
        .replace('_', " ")
        .replace('-', " ");
    normalize_text(&stem)
        .map(title_case)
        .unwrap_or_else(|| format!("Chapter {}", index + 1))
}

fn title_case(value: String) -> String {
    value
        .split_whitespace()
        .map(|word| {
            let mut chars = word.chars();
            match chars.next() {
                Some(first) => format!("{}{}", first.to_uppercase(), chars.as_str()),
                None => String::new(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

fn zip_dirname(path: &str) -> String {
    path.rsplit_once('/')
        .map(|(dir, _)| dir.to_string())
        .unwrap_or_default()
}

fn join_zip_path(base: &str, path: &str) -> String {
    if base.is_empty() {
        normalize_zip_path(path)
    } else {
        normalize_zip_path(&format!("{base}/{path}"))
    }
}

fn strip_base(path: &str, base: &str) -> String {
    if base.is_empty() {
        normalize_zip_path(path)
    } else {
        path.strip_prefix(&format!("{base}/"))
            .map(normalize_zip_path)
            .unwrap_or_else(|| normalize_zip_path(path))
    }
}

fn normalize_zip_path(path: &str) -> String {
    let mut parts = Vec::new();
    let normalized = path.replace('\\', "/");
    for part in normalized.split('/') {
        match part {
            "" | "." => {}
            ".." => {
                parts.pop();
            }
            value => parts.push(value),
        }
    }
    parts.join("/")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_zip_paths() {
        assert_eq!(
            normalize_zip_path("OPS/Text/../Images/cover.jpg"),
            "OPS/Images/cover.jpg"
        );
        assert_eq!(
            normalize_zip_path("./chapter\\one.xhtml"),
            "chapter/one.xhtml"
        );
    }

    #[test]
    fn parses_readable_paragraph_blocks_without_duplicate_wrapper_divs() {
        let blocks = parse_chapter_blocks(
            r#"
            <html><body>
              <div><p>First paragraph.</p><p>Second paragraph.</p></div>
              <div>Loose paragraph in a div.</div>
            </body></html>
            "#,
            "",
        );
        let texts = blocks
            .into_iter()
            .map(|block| block.text)
            .collect::<Vec<_>>();
        assert_eq!(
            texts,
            vec![
                "First paragraph.",
                "Second paragraph.",
                "Loose paragraph in a div."
            ]
        );
    }

    #[test]
    fn skips_deep_wrapper_divs_around_chapter_content() {
        let blocks = parse_chapter_blocks(
            r#"
            <html><body>
              <div class="galley"><section><h1>4</h1><p>First paragraph.</p><p>Second paragraph.</p></section></div>
            </body></html>
            "#,
            "",
        );
        let texts = blocks
            .into_iter()
            .map(|block| block.text)
            .collect::<Vec<_>>();
        assert_eq!(texts, vec!["4", "First paragraph.", "Second paragraph."]);
    }

    #[test]
    fn resolves_image_paths_against_the_chapter_directory() {
        let blocks = parse_chapter_blocks(
            r#"<html><body><img src="../Images/plate01.jpg" alt="Plate"></body></html>"#,
            "OPS/Text",
        );

        assert_eq!(blocks.len(), 1);
        assert!(matches!(blocks[0].kind, ExtractedBlockKind::Image));
        assert_eq!(
            blocks[0].asset_path.as_deref(),
            Some("OPS/Images/plate01.jpg")
        );
        assert_eq!(blocks[0].alt, "Plate");
    }
}
